scripts/extensions/sams/src/components/sets/setEditorPanel.tsx

```diff
@@ -0,0 +1,245 @@
+    FormGroup,
+    FormRow,
+    PanelContent,
+    PanelContentBlock,
+    PanelContentBlockInner,
+    PanelHeader,
+    PanelHeaderSlidingToolbar,
+    Text,
+} from '../../ui';
+
+// Utils
+import {hasItemChanged} from '../../utils/api';
+
+interface IProps {
+    original?: ISetItem;
+    destinations: Array<IStorageDestinationItem>;
+    closeEditor(): void;
+    previewSet(set: ISetItem): void;
+    updateSet(original: ISetItem, updates: Partial<ISetItem>): Promise<ISetItem>;
+    createSet(set: Partial<ISetItem>): Promise<ISetItem>;
+    currentDestination?: IStorageDestinationItem;
+}
+
+interface IState {
+    updates: Partial<ISetItem>;
+    isDirty?: boolean;
+}
+
+export function getSetEditorPanel(superdesk: ISuperdesk) {
+    const {gettext} = superdesk.localization;
+
+    const mapStateToProps = (state: IApplicationState) => ({
+        original: getSelectedSet(state),
+        destinations: getStorageDestinations(state),
+        currentDestination: getSelectedSetStorageDestination(state),
+    });
+
+    const mapDispatchToProps = (dispatch: Dispatch) => ({
+        closeEditor: () => dispatch(closeSetContentPanel()),
+        previewSet: (set: ISetItem) => dispatch(previewSet(set._id)),
+        updateSet: (original: ISetItem, updates: ISetItem) => dispatch<any>(updateSet(original, updates)),
+        createSet: (set: ISetItem) => dispatch<any>(createSet(set)),
+    });
+
+    class SetEditorPanelComponent extends React.Component<IProps, IState> {
+        onChange: Dictionary<string, (value: any) => void>;
+
+        constructor(props: IProps) {
+            super(props);
+
+            if (this.props.original?._id == null) {
+                this.state = {
+                    updates: {
+                        destination_name: this.props.destinations?.[0]?._id,
+                    },
+                    isDirty: true,
+                };
+            } else {
+                this.state = {
+                    updates: cloneDeep<ISetItem>(this.props.original),
+                    isDirty: false,
+                };
+            }
+
+            this.onStateChange = this.onStateChange.bind(this);
+            this.onSave = this.onSave.bind(this);
+            this.onCancel = this.onCancel.bind(this);
+
+            this.onChange = {
+                name: (value: string) => this.onFieldChange('name', value),
+                description: (value: string) => this.onFieldChange('description', value),
+                destination_name: (value: string) => this.onFieldChange('destination_name', value),
+                state: (value: boolean) => this.onStateChange(value),
+            };
+        }
+
+        onFieldChange(field: keyof ISetItem, value: any): void {
+            const updates = this.state.updates;
+            let dirty = true;
+
+            updates[field] = value;
+
+            if (this.props.original != null) {
+                dirty = hasItemChanged<ISetItem>(this.props.original, this.state.updates);
```

Comment:
It's not wrong, but type argument doesn't provide any additional constraints if types of arguments are specified